### PR TITLE
Move injectors code to external library

### PR DESCRIPTION
**Pull request by BenLorantfy**
This PR moves the injector-related code to a new external library called `redux-injectors`. For the full discussion on how we got here, see [this issue](https://github.com/react-boilerplate/react-boilerplate/issues/2698). A few guiding notes:
- See this file for the changes made to the store setup: https://github.com/react-boilerplate/react-boilerplate/pull/2749/files#diff-ea94aca90fbef7f544d2a96cf1da5f8e
   - Removes the store mutation in favour of using a [store enhancer](https://redux.js.org/recipes/configuring-your-store#extending-redux-functionality)
   - Store is now given `createReducer`. It was only given `runSaga` previously. 
- This is an example of how the imports would change: https://github.com/react-boilerplate/react-boilerplate/pull/2749/files#diff-b40dc3de2b66cf46597584200a5599f7R14
- Removed a bunch of code:
  - Removed the injectReducer, injectSaga, checkStore, and sagaInjectors files
  - Removed the related tests
  - Removed the same files in the `templates` folder
- Added `docs/js/redux-injectors.md`. I linked to the `redux-injectors` docs for readers to view the full docs, but let me know if I should go into more detail in this file.
- Changed a few documentation files to use `redux-injectors` (e.g. `docs/js/redux-saga.md`, `docs/js/redux.md`)
- Removed some steps from `docs/js/remove.md` 
- Updated the generators to use `redux-injectors`
- Updated all the templates with the same changes that were done to the source code

A few other links:
- redux-injectors on npm: https://www.npmjs.com/package/redux-injectors
- redux-injectors on GitHub: https://github.com/react-boilerplate/redux-injectors

### CODE REVIEW

**Comment by julienben** [comment]
😍😍😍
image

Quick question: Why do we need to add aliases for react deps in the webpack config?

**Comment by BenLorantfy** [comment]
That was just for local development and is needed when using `yarn link` to link `redux-injectors` to `react-boilerplate`. Planning on removing it once `redux-injectors` is published and before merging this.

**Comment by BenLorantfy** [comment]
OK, this is now ready for review whenever anyone has a chance 🙂

**Comment by BenLorantfy** [comment]
@julienben When you have some time, I also need help transferring redux-injectors to the reactboilerplate org as I explained in this comment: https://github.com/react-boilerplate/react-boilerplate/issues/2698#issuecomment-536151715

**Comment by julienben** [comment]
I think the transfer is done:
image
Will review the code asap. Please notice that CI isn't passing though. I think it's related to coverage having dropped.

**Comment by BenLorantfy** [comment]
Oh, I think it dropped because I removed a lot of code that was well tested, I'll see if I can add some more tests elsewhere.

**Comment by julienben** [comment]
I haven't taken a deep look at this PR's changes yet but maybe also check if some of the tests you removed were testing code **outside** the injectors?

**Comment by BenLorantfy** [comment]
I checked and I don't think I removed any tests that were testing other code. I did remove some `configureStore` tests but the coverage for that file is still 100%.

I fixed the coverage by adding some tests for LanguageProvider and Loadable. Also copied the folder `app/containers/LanguageProvider/tests` to `internals/templates/containers/LanguageProvider/tests` because it was missing previously.

**Comment by julienben** [comment]
Thanks!!

Just one thing: Can you undo the copying of `app/containers/LanguageProvider/tests`? This is already handled by the cleaning script (lines 17 to 20).

**Comment by BenLorantfy** [comment]
Ah I see, done

File: app/configureStore.js [hunk 1]
```diff
@@ -28,30 +29,32 @@ export default function configureStore(initialState = {}, history) {
   }
 
   const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions);
+  const runSaga = sagaMiddleware.run;
```
Comment (julienben) [inline app/configureStore.js]: Slightly prettier maybe:
```js
const { run: runSaga } = sagaMiddleware;
```

File: app/containers/LanguageProvider/tests/selectors.test.js [hunk 2]
```diff
@@ -1,11 +1,17 @@
...
+    const languageState = {};
     const mockedState = {
-      language: globalState,
+      language: languageState,
     };
-    expect(selectLanguage(mockedState)).toEqual(globalState);
+    expect(selectLanguage(mockedState)).toEqual(languageState);
+  });
+
+  it('should select the initial state when state is missing', () => {
```
Comment (julienben) [inline app/containers/LanguageProvider/tests/selectors.test.js:13]: 👌

File: app/tests/store.test.js [hunk 3]
```diff
@@ -6,28 +6,9 @@ import { browserHistory } from 'react-router-dom';
...
-      expect(typeof store.injectedSagas).toBe('object');
-    });
-  });
-
-  describe('runSaga', () => {
-    it('should contain a hook for `sagaMiddleware.run`', () => {
-      expect(typeof store.runSaga).toBe('function');
-    });
+  it('should return a redux store', () => {
+    const store = configureStore({}, browserHistory);
```
Comment (julienben) [inline app/tests/store.test.js]: Here browserHistory is undefined because it's an old obsolete API from react-router. You'll see it's been removed in #2791. You'll get a merge conflict here. You can just remove that argument and the test should pass just fine without it.

File: app/tests/store.test.js [hunk 4]
```diff
@@ -6,28 +6,9 @@ import { browserHistory } from 'react-router-dom';
...
+    expect(typeof store).toBe('object');
```
Comment (julienben) [inline app/tests/store.test.js]: It's a bit light to say it should return a redux store but then just check that it's of type object. You should use some of the same checks that used to be in checkStore. And IMO there's nothing wrong with also checking for the presence of the keys which are added by redux-injectors (runSaga, injectedSagas, injectedReducers, and replaceReducer too or is that gone now?).

File: app/utils/tests/loadable.test.js [hunk 5]
```diff
@@ -18,6 +22,13 @@ describe('loadable', () => {
     expect(firstChild).toMatchSnapshot();
   });
 
+  it('should render null by default with empty options', () => {
```
Comment (julienben) [inline app/utils/tests/loadable.test.js:25]: 👌

File: docs/js/remove.md [hunk 6]
```diff
@@ -10,31 +10,8 @@ If you really want to get rid of it, you will have to remove its presence from s
...
-**app/utils/tests**
-
-1.  Remove two files: `injectSaga.test.js` and `sagaInjectors.test.js`
-
-**app/utils/tests/checkStore.test.js**
-
-1.  Remove `expect(() => checkStore({ ...store, injectedSagas: null })).toThrow();`
-2.  Remove `expect(() => checkStore({ ...store, runSaga: null })).toThrow();`
+4.  Remove statement `const runSaga = sagaMiddleware.run;`
+5.  Remove `runSaga` from `createInjectorsEnhancer` params
```
Comment (julienben) [inline docs/js/remove.md:14]: Does this actually work? You can use redux-injectors without redux-saga?

File: docs/js/redux-injectors.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
+# `redux-injectors`
+
+`redux-injectors` is a library we made ourselves by moving code we used to have
+in the utils folder into a seperate library. It allows you to dynamically load
```
Comment (julienben) [inline docs/js/redux-injectors.md]: I would replace the first sentence with something like this:

> [`redux-injectors`](https://github.com/react-boilerplate/redux-injectors) is an official `react-boilerplate` companion library. We built it so that it can be used and maintained independently from `react-boilerplate`.
> 
> It allows you to dynamically load...

It feels slightly more professional IMO.

File: internals/generators/container/index.js.hbs [hunk 8]
```diff
@@ -21,11 +21,18 @@ import { useSelector, useDispatch } from 'react-redux';
...
+  {{else}}
+import { useInjectReducer } from 'redux-injectors';
+  {{/if}}
+{{else}}
+  {{#if wantSaga}}
+import { useInjectSaga } from 'redux-injectors';
+  {{/if}}
 {{/if}}
 {{#if wantActionsAndReducer}}
-import { useInjectReducer } from 'utils/injectReducer';
```
Comment (julienben) [inline internals/generators/container/index.js.hbs:28]: Do you mind opening a separate PR if you'd like to update the logic for the container generator?

(I agree that there's an issue if `wantActionsAndReducer` is false but `wantSaga` is true but it should be handled separately from this.)

Comment (julienben) [reply internals/generators/container/index.js.hbs:28]: By this I mean please only update the imports here.

File: docs/js/redux-injectors.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
+# `redux-injectors`
```
Comment (julienben) [inline docs/js/redux-injectors.md:1]: Are you linking to this file from anywhere else in the docs?

**Review by julienben — CHANGES_REQUESTED** [review]
Hey! Looks awesome overall. Just left a few small comments.

Also:
- Make sure you merge the latest changes from dev and fix the conflicts. (Sorry, I merged a few PRs over the weekend 😅)
- Did you check if code-splitting/loadable still works the same? I'm just worried that we lost that feature partially since the containers import from `redux-injectors` and the store setup does too. To test it, compare the webpack output of a build from the dev branch and your branch. They won't be exactly the same but you should see one file for each container. (Use webpack analyze if necessary.)
- Any comments on files in the app folder also apply to the internals of course.

One more thing you might want to do is run a regression test as described in `deployment.md`. Just make sure you uncomment the `git commit` line in `clean.js`.

File: docs/js/redux-injectors.md (hunk 9 above)
Comment (julienben) [reply docs/js/redux-injectors.md:1]: You might want to or very few people might ever read it.

File: docs/js/remove.md (hunk 6 above)
Comment (BenLorantfy) [reply docs/js/remove.md:14]: kind of. It throws some invariant warnings. I could probably disable them in `redux-injectors`

File: internals/generators/container/index.js.hbs (hunk 8 above)
Comment (BenLorantfy) [reply internals/generators/container/index.js.hbs:28]: I could just change the import but then you could up with this:
```
import { useInjectReducer } from 'redux-injectors';
import { useInjectSaga } from 'redux-injectors';
```

Which I'm pretty sure will throw a linting error under https://eslint.org/docs/rules/no-duplicate-imports , and it seems bad to have the generator generate a file with a linting error.

File: docs/js/redux-injectors.md (hunk 9 above)
Comment (BenLorantfy) [reply docs/js/redux-injectors.md:1]: Linked to it from `js/README.md`

File: app/tests/store.test.js (hunk 4 above)
Comment (BenLorantfy) [reply app/tests/store.test.js]: So I added in a more robust test for checking if it returned a redux store. I'm hesitant to check for `injectedSagas`, `injectedReducers`, etc, because I'd consider that an internal implementation detail of `redux-injectors`.  I'd like to be able to change that in the future if we need to without breaking `react-boilerplate` tests.

**Comment by BenLorantfy** [comment]
Looks like code-splitting is still working, I can see `FeaturePage` and `HomePage` are in two different files in the webpack output:

Screen Shot 2019-10-27 at 1 13 00 PM

### Regression test:

**Example App:**
- [x] Browse example app on development server
  - [x] Browse Features page, change language to `de`
  - [x] Browse NotFound page
- [x] Browse example app on dev tunnel
- [x] Browse example app on Production server
- [x] Browse example app offline

**Internal Commands:**

- [x] `npm run clean` (Be careful, this will commit all your changes.)
- [x] `npm run generate component` TestComp /w defaults
- [x] `npm run generate container` TestPage /w defaults

File: docs/js/remove.md (hunk 6 above)
Comment (julienben) [reply docs/js/remove.md:14]: Could it be done by checking which keys are passed to createInjectorsEnhancer? It can be like a configuration object for the injectors and will be especially useful later if/when you add more integrations to them.

Comment (julienben) [reply docs/js/remove.md:14]: (No emergency here. We can merge this branch without this being present in injectors.)

File: docs/js/redux-saga.md [hunk 10]
```diff
@@ -49,19 +49,18 @@ export default function* rootSaga() {
 Then, in your `index.js`, use our custom hook to inject the root saga:
```
Comment (julienben) [inline docs/js/redux-saga.md]: This text here says "our custom hook". Could be updated as well no?

File: docs/js/README.md [hunk 11]
```diff
@@ -23,6 +23,7 @@ add new parts of your application!
 - [Immer](immer.md)
 - [reselect](reselect.md)
 - [redux-saga](redux-saga.md)
+- [redux-injectors](redux-injectors.md)
```
Comment (julienben) [inline docs/js/README.md:26]: 👌👌

File: docs/js/redux-saga.md (hunk 10 above)
Comment (julienben) [reply docs/js/redux-saga.md]: (with an additional link to your redux-injectors.md file for example)

File: internals/generators/container/index.js.hbs (hunk 8 above)
Comment (julienben) [reply internals/generators/container/index.js.hbs:28]: Oh my bad! I went over this too fast. The logic remains the same but it looks more complicated because you're importing from one package instead of 2 files.

File: internals/scripts/helpers/get-npm-config.js [hunk 12]
```diff
@@ -0,0 +1,3 @@
+const fs = require('fs');
+
+module.exports = JSON.parse(fs.readFileSync('package.json', 'utf8'));
```
Comment (julienben) [inline internals/scripts/helpers/get-npm-config.js]: Merge conflict issue. This file shouldn't be here.

File: internals/templates/tests/i18n.test.js [hunk 13]
```diff
@@ -0,0 +1,28 @@
+import { formatTranslationMessages } from '../i18n';
```
Comment (julienben) [inline internals/templates/tests/i18n.test.js]: This file shouldn't be here no?

File: internals/scripts/npmcheckversion.js [hunk 14]
```diff
@@ -0,0 +1,8 @@
+const { exec } = require('child_process');
```
Comment (julienben) [inline internals/scripts/npmcheckversion.js]: This file shouldn't be here either.

**Review by julienben — CHANGES_REQUESTED** [review]

File: docs/js/remove.md (hunk 6 above)
Comment (BenLorantfy) [reply docs/js/remove.md:14]: Possibly, I'd have to look into it more

File: internals/scripts/helpers/get-npm-config.js (hunk 12 above)
Comment (BenLorantfy) [reply internals/scripts/helpers/get-npm-config.js]: ugh not sure how that happened

File: docs/js/remove.md (hunk 6 above)
Comment (julienben) [reply docs/js/remove.md:14]: No rush. Just consider it for a future release of injectors.

**Review by julienben — APPROVED** [review]
@BenLorantfy it's been a pleasure working with you on this. Thanks for your patience and your attention to detail! I hope you'll continue contributing!

**Comment by BenLorantfy** [comment]
No problem, thanks for your guidance and help with this 👍
